### PR TITLE
statically compress webpack artifacts

### LINKED ISSUE #4285: [Webpack] Assets could be compressed to improve delivery times
Problem with current Theia is when there is no CDN to back it up, `bundle.js` is slow to serve - causing the browser to sometimes drop the connection before it even completes.

I suggest to run compression on the assets so we can improve delivery times.

Webpack has a dedicated [plugin](https://webpack.js.org/plugins/compression-webpack-plugin/) for this.

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- [x] fix #4285: enable compressions of Theia build artifacts, i.e. bundles and monaco scripts, styles, wasm and so on.
- [x] https://github.com/webpack-contrib/compression-webpack-plugin/issues/133 - in dev mode compressed artifacts are not reused, which leads to redownloading of all bundles on each refresh
  -  We have to decide whether we disable it for now in dev mode. If we do when for the initial load we have to pull 22.5 mb instead of 4.8mb. If we don't then during consequent incremental development we have to pull around 2Mb each time instead of only rebuilt bundles as before, e.g. for a change in the plugin ext it is 0.8Mb. cc @svenefftinge @AlexTugarev 
- [x] check license compatibility of new dependencies

##### Production mode

- before 9.3Mb is transferred:  
before-prod

- after 2.5Mb is transferred:
after-prod

##### Development mode

- before 22.5Mb is transferred:
before-dev

- after 4.8Mb is transferred:
after-dev

#### How to test

- Build in production and development before and after changes.
- Load the application without the cache and check how much traffic is transferred in each time.
- In development mode pay attention whether bundles are reused on incremental change in extensions.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2624,6 +2629,26 @@ cacache@^10.0.4:
     unique-filename "^1.1.0"
     y18n "^4.0.0"
 
+cacache@^11.2.0:
```
Comment (akosyakov) [inline yarn.lock]: @marcdumais-work could you run license check please?

**Comment by benoitf** [comment]
I would not use it for dev mode.

File: yarn.lock (hunk 1 above)
Comment (marcdumais-work) [reply yarn.lock]: For you Anton, sure :) 
I'll report back.

Comment (marcdumais-work) [reply yarn.lock]: I ran the license check and everything looks good

**Comment by akosyakov** [comment]
> I would not use it for dev mode.

Yeah, in such state I agree, but generally i would like to have it for the dev mode. I had bad experience of developing Theia from airports WiFis, Russian on mobile connections and obviously with EclipseCon WiFi in Ludwisburg. Looking into fixing compression with watch mode in the webpack plugin.

**Comment by akosyakov** [comment]
I've opened a PR against the compression plugin to improve the development mode: https://github.com/webpack-contrib/compression-webpack-plugin/pull/134

For time being, I've published our fork with a fix and used it for a PR.

Results are pretty good, now in the development mode a change in the plugin system requires redownloading 134 kb vs 800kb before.

Screen Shot 2019-10-14 at 16 47 08

And initially one has to download 4.6 Mb for development instead of 22.5 before.

Screen Shot 2019-10-14 at 16 46 33

**Comment by akosyakov** [comment]
@AlexTugarev could you review, confirm results please? It would be good to have it done before EclipseCon.

**Comment by akosyakov** [comment]
I will add an option to pass to theia webpack to disable it.

**Comment by akosyakov** [comment]
Clarifications how to test:
- disable cache in network tab
- clean network tab and logs
- reload the page
- check **transfered** data in the network tab
- test this PR and build from master to see the difference

**Comment by AlexTugarev** [comment]
this PR

Screen Shot 2019-10-15 at 09 58 31

vs current master

Screen Shot 2019-10-15 at 09 59 52

**Comment by benoitf** [comment]
we won't merge it until it PR is merged upstream, correct ?

**Comment by akosyakov** [comment]
@AlexTugarev yes, it shows you that for this PR 4.6 Mb is transferred vs 23.9 Mb on master. You should also check in dev mode, i.e.
- disable `disable cache`
- watch example app
- watch some extension
- change some extension
- check how much is loaded against master

**Comment by akosyakov** [comment]
> we won't merge it until it PR is merged upstream, correct ?

Why? We tested that it works for us in different conditions. PR in upstream has tests as well. webpack guys can review it for several months, EclipesCon is next wekk.

**Comment by benoitf** [comment]
But if we do forks on each issue we had on upstream projects it's like nightmare to maintain.
I don't see why we should use code from un-merged PR upstream.

>  EclipesCon is next wekk.

About EclipseCon, I don't see how it's related. If you work locally you don't transfer new stuff each time. If it's about hosted version, it's not related to Eclipse Theia but more a theia/gitpod assembly issue.

**Comment by AlexTugarev** [comment]
> If it's about hosted version, it's not related to Eclipse Theia but more a theia/gitpod assembly issue.

@benoitf we want to make a Theia workshop, and this PR will improve the development experience a lot.

**Comment by akosyakov** [comment]
> About EclipseCon, I don't see how it's related. If you work locally you don't transfer new stuff each time. If it's about hosted version, it's not related to Eclipse Theia but more a theia/gitpod assembly issue.

Yes, we are going to use Gitpod, but it is not related to how we deploy Gitpod. It is about improve development of Theia remotely. Gitpod, or any other services, should not do any caching of user apps. I've added to dev meeting. It is important for good experience next week. cc @marcdumais-work @svenefftinge

**Comment by akosyakov** [comment]
> But if we do forks on each issue we had on upstream projects it's like nightmare to maintain.
I don't see why we should use code from un-merged PR upstream.

It is not ideal, but sometimes you should move forward somehow. We do it for `@theia/node-pty` and `@typefox/monaco-editor-core`.

**Comment by benoitf** [comment]
well, IMHO we shouldn't

For example node-pty upstream is 0.8.1, forked is still on 0.7.8
and monaco editor core is 0.18.1 while fork is 0.18.0-next.1 (so you don't know well the diff as well)

published package are also referencing to the upstream github URLs and not the forked one.

also I don't see why sometimes we use `@theia` namespace and sometimes `@typefox`

**Comment by akosyakov** [comment]
For Monaco there is no another way. We want to consume internal APIs of VS Code editor to implement VS Code support. Default Monaco distribution tree shakes a lot making it impossible.

For other, I'm agree that we should move to latest version at some point. But It's decided by a need of a project, not by desire to use latest versions till then it does not cause any issues. Otherwise we will spend all our time updating dependencies each day. If there is no important bug fixes and improvements in these dependencies. We can live with our forks forever there is no a tragedy here.

**Comment by akosyakov** [comment]
> also I don't see why sometimes we use @theia namespace and sometimes @typefox

That's more historical reason. I will republish compression plugin under @theia.

**Comment by akosyakov** [comment]
> published package are also referencing to the upstream github URLs and not the forked one.

I think it is good. Bugs should be opened against the parent repo. That's also for example that devs do who forked monaco-languageclient last year. They send users to the main repo. It's also not really important since only we use it.

**Comment by akosyakov** [comment]
- tests should be fixed
- one can pass `--no-static-compression` to `theia build` or `theia watch` to disable it
  - the record added to CHANGELOG
- the fork moved from my account under theia-org: https://github.com/theia-ide/compression-webpack-plugin/tree/theia_publishing
  - the package is published under `@theia` npm scope as well

@AlexTugarev have you managed to reproduce perf improvements in dev mode?

**Comment by akosyakov** [comment]
We've agreed during the dev meeting with @marcdumais-work :
- that this PR improves production experience generally, especially for our docker images;
- it improves development mode on slow network;
- that using our fork is fine whenever we find is practical, although:
  - our solution should be properly communicated to the upstream project as an issue or a PR;
  - fork should be under theia-ide org, under `@theia` npm scope name;
  - there should be an issue to switch to upstream when an issue gets resolve there.

We can transfer monaco-editor-core under theia-ide org during next migration: https://github.com/eclipse-theia/theia/issues/6223#issuecomment-542245435 Getting rid of Monaco fork is impossible, since it makes the plugin system impossible to implement.

@AlexTugarev Could you finish the review please?

File: dev-packages/application-manager/src/generator/webpack-generator.ts [hunk 2]
```diff
@@ -38,19 +38,50 @@ const webpack = require('webpack');
...
 const CircularDependencyPlugin = require('circular-dependency-plugin');
+const CompressionPlugin = require('@theia/compression-webpack-plugin')
 
 const outputPath = path.resolve(__dirname, 'lib');
-const { mode }  = yargs.option('mode', {
+const { mode, staticCompression }  = yargs.option('mode', {
     description: "Mode to use",
     choices: ["development", "production"],
     default: "production"
+}).option('static-compression', {
```
Comment (AlexTugarev) [inline dev-packages/application-manager/src/generator/webpack-generator.ts]: q: `no-static-compression` vs `static-compression`?

Comment (AlexTugarev) [reply dev-packages/application-manager/src/generator/webpack-generator.ts]: https://github.com/yargs/yargs/blob/master/docs/tricks.md#negate

**Pull request merged**